docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+
+The design presented here requires:
+
+*   A durable, ordered, and immutable commit log.
+*   A "local" storage mechanism which can support the operations required by
+    the Trillian log_storage API.
+
+
+
+## Design Overview
+
+![overview diagram](commit_log_based_storage_design_overview.png "Overview")
+
+The `leaves` topic is the canonical source of truth for the ordering of leaves
+in a log.
+
+The `STHs` topic is a list of all STHs for a given log.
+
+Kafka topics are configured never to expire entries (this is a supported mode),
+and Kafka is known to scale to multiple terabytes within a single partition.
+
+HBase instances are one-per-cluster, built from the contents of the Kafka
+topics, and, consequently, are essentially disposable.
+
+Queued leaves are sent by the Trillian frontends to the Kafka `Leaves` topic.
+Since Kafka topics are append-only and immutable, this effectively sequences
+the entries in the queue.
+The current master sequencer consumes the `Leaves` topic, uses the Kafka sequence
+number of each dequeued leaf as its Trillian Log sequence number, and adds it to
+the local Merkle tree ensuring the sequence numbers match.
+
+When a new tree head is created, the sequencer writes this to the STHs topic
+before committing the tree update to the local database.
+
+Since the commit log forms the source of truth for the log entry ordering and
+committed STHs, everything else can be derived from that. This means that
+updates to the serving HBase DBs can be made to be idempotent, which means that
+the transactional requirements of Trillian's LogStorage APIs can be relaxed.
+
+The addition of this style of storage implementation requires that Trillian
+does not guarantee the perfect deduplication of entries, even though it may be
+possible to do so with some storage implementations.  i.e. personalities MUST
+present LeafIdentityHashes, and Trillian MAY deduplicate.
+
+## Detailed Design
+
+#### Enqueuing leaves
+
+RPC calls to frontend `QueueLeaves` results in the leaves being individually
+added to the Kafka topic `Leaves`.  They need to be added individually to allow
+the Kafka topic sequencing to be the definitive source of log sequence
+information.
+
+Log frontends may attempt to de-duplicate incoming leaves by consulting the
+local storage DB using the identity hash, but this will always be a "best
+effort" affair, so the Trillian APIs must not assume that duplicates are
+impossible, even though in practice, when using other storage
+implementations, they may well be so currently.
+
+#### Master election
+
+Multiple sequencers may be running to provide resilience, if this is the case
+there must be a mechanism for choosing a single master instance among the
+running sequencers.
+
+The Trillian repo provides an etcd-backed implementation of this already.
+
+A sequencer must only participate/remain the master if its local database state
+is at least as new at the latest message in the Kafka `STHs` topic.
```

Comment:
Sounds like we might need special handling for an empty tree? A dummy zero STH should work like we currently have?